### PR TITLE
fix: detach reward passed to MSTDP learners to prevent memory growth

### PR DESCRIPTION
**Pull request by JPL11**
## What

Detach the external `reward` at the top of `MSTDPLearner.step()` and `MSTDPETLearner.step()`, so a graph-connected reward can never leak the forward pass's autograd graph into `synapse.weight.grad` or the returned `delta_w`. Adds three regression tests and a CHANGELOG entry.

Follow-up to #703; this closes the remaining leak path noted in that PR's description ("a graph-connected `reward` could still leak via `dw = reward * trace_e`").

## Why

#703 detached the spikes the learners *record*, but the reward the caller *passes in* still flows straight into the weight update: `dw = (reward.view(-1, 1, 1) * self.eligibility).sum(0)` (MSTDP) and `dw = reward * self.trace_e` (MSTDPET). In reward-modulated training the reward is often the output of a differentiable module (e.g. a critic network). If the caller doesn't detach it, `step(on_grad=True)` writes a graph-connected tensor into `weight.grad`, and each subsequent `step()` chains onto it via `weight.grad - delta_w` — exactly the retention pattern of #576, just entering through the reward instead of the spikes.

Measured (CPU, torch 2.13, Linear 8→5, critic `nn.Linear(5, 1)`, 20 steps per run, model re-created per run): **before** — `weight.grad.grad_fn` is populated and +122 tensors survive `gc.collect()` per run, growing linearly; **after** — `grad_fn is None`, 0 surviving tensors.

The same reasoning as #703 applies to why detaching is safe: STDP-family updates already operate through `weight.data`/`weight.grad` and `step()` is not a supported differentiable path, so no gradient route is removed. Backpropagating into the critic itself is unaffected — users do that from their own loss on the critic's output, not through the learner's weight update.

## Verification

- `test_mstdp_learners_step_detaches_reward`, `test_mstdp_learner_returns_detached_delta_w`, and `test_mstdp_learners_free_tensors_with_graph_connected_reward` all fail on master and pass with this fix.
- Full suite: 1347 passed, 210 skipped (CPU) — no regressions.
- `uv format --check` clean on the changed files; `docs/source/changelog.rst` regenerated via `tools/generate_changelog_rst.py`.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed MSTDPLearner and MSTDPETLearner behavior to detach tensor rewards during `step`, preventing autograd graph leakage into weight gradients and returned updates.
  * Resolved autograd graph retention that could cause memory growth across training iterations.
* **Tests**
  * Added new regression tests covering critic-connected reward behavior, ensuring gradients/returned `delta_w` are detached when expected.
  * Added repeated-run checks to confirm tensor cleanup does not degrade over time.

### CODE REVIEW

File: test/activation_based/test_learning.py [hunk 1]
```diff
@@ -173,10 +173,111 @@ def test_mstdp_learners_records_are_detached():
...
+    out_spike = sn(fc(in_spike))
+    reward = critic(out_spike).squeeze(-1)
+
+    delta_w = learner.step(reward, on_grad=False)
+    assert delta_w.grad_fn is None
+    assert not delta_w.requires_grad
+
+    functional.reset_net(sn)
+    learner.reset()
+
```
[the remark this answers is not in the document]
Comment (JPL11) [reply test/activation_based/test_learning.py]: Good catch — extended the test to loop over both learners (same pattern as `test_mstdp_learners_records_are_detached`), with unbatched spikes and a scalar reward for the MSTDPET case. Done in 222f36c9.

File: test/activation_based/test_learning.py [hunk 2]
```diff
@@ -173,10 +173,111 @@ def test_mstdp_learners_records_are_detached():
...
+
 if __name__ == "__main__":
     test_stdp_learner_records_are_detached()
     test_stdp_learner_step_does_not_retain_graph()
     test_stdp_learner_matches_functional_update()
     test_stdp_learner_frees_tensors_across_runs()
     test_mstdp_learners_records_are_detached()
+    test_mstdp_learners_step_detaches_reward()
+    test_mstdp_learner_returns_detached_delta_w()
```
[the remark this answers is not in the document]
Comment (JPL11) [reply test/activation_based/test_learning.py]: Done in 222f36c9.

**Review by AllenYolk — DISMISSED** [review]
The implementation and regression coverage look correct, and the targeted test file passes locally (8 passed). One repository-standard documentation update is still required before merge.

Both `MSTDPLearner.step()` and `MSTDPETLearner.step()` are public APIs whose behavior now changes: tensor rewards are detached inside `step()`, so neither `weight.grad` nor a returned `delta_w` remains connected to the reward autograd graph. Please document this behavior in both the Chinese and English sections of each method docstring (including the `reward`/return semantics as appropriate), keeping the two languages equivalent.

I also verified `uv format --check -- spikingjelly/activation_based/learning.py test/activation_based/test_learning.py` and `uv run python tools/generate_changelog_rst.py --check`.

**Review by AllenYolk — APPROVED** [review]
The implementation and regression coverage look good. I verified the targeted learning tests (8 passed), changed-file formatting, and changelog generation locally.

The bilingual `step()` docstrings could additionally state that tensor rewards are detached from autograd; maintainers will handle that small documentation clarification as a follow-up, so it is not blocking this PR.
